integrations/datadog/client.py

```diff
@@ -233,6 +296,117 @@ async def get_slo_history(
+        """Extracts the Datadog metric name from a query string.
+
+        Args:
+            query_string: The Datadog query string.
+
+        Returns:
+            The metric name, or None if no metric name is found.
+
+        Examples:
+            - "sum:container.memory.usage" -> "container.memory.usage"
+            - "avg:system.cpu.used{service:my-service}" -> "system.cpu.used"
+        """
+        match = re.search(r"(?:\w+:)?(\w+\.\w+\.\w+)", query_string)
+        if match:
+            return match.group(1)
+        else:
+            return None
+
+    async def get_single_service(self, service_id: str) -> dict[str, Any] | None:
+        if not service_id:
+            return None
+        url = f"{self.api_url}/api/v2/services/definitions/{service_id}"
+        return await self._send_api_request(url)
+
+    async def get_metric_metadata(self, metric: str) -> dict[str, Any] | None:
+        url = f"{self.api_url}/api/v1/metrics/{metric}"
+        return await self._send_api_request(url)
+
+    async def get_metrics(
+        self, query: str, env: str = "prod", service: str = "*"
+    ) -> AsyncGenerator[list[dict[str, Any]], None]:
+        """
+        Fetches metrics for specified services and environment.
+
+        Args:
+            query (str): The Datadog query string to execute.
+            env (str): The environment to filter by (default: "prod").
+            service (str): The service ID to filter by, or "*" to fetch metrics for all services.
+
+        Yields:
+            AsyncGenerator[list[dict[str, Any]], None]: A list of metrics for each service.
+        """
+        url = f"{self.api_url}/api/v1/metrics"
+
+        metrics = []
+
+        services = []
+        if service == "*":
+            # get all services
+            async for service_list in self.get_services():
+                services.extend(service_list)
+        else:
+            # get service by id
+            result = await self.get_single_service(service)
+            if not result.get("data"):
+                logger.warning(f"Service with id {service} not found")
+                return
+
+            services = [result.get("data")]
+
+        logger.info(f"Fetching metrics for {len(services)} services")
+
+        for service in services:
+            service_id = service["attributes"]["schema"]["dd-service"]
+
+            params = {"env": env, "service": service_id}
+            query_with_values = self._create_query_with_values(
+                f"{query}{{service:{service_id},env:{env}}}", params
+            )
+
+            time_in_minutes = 10
+            end_time = int(time.time())
+            start_time = end_time - (time_in_minutes * 60)
+
+            url = f"{self.api_url}/api/v1/query?from={start_time}&to={end_time}&query={query_with_values}"
+            result = await self._fetch_with_rate_limit_handling(
+                url, semaphore=self._metrics_semaphore
+            )
+
```

Comment:
you are not even running it concurrently how come you need the semaphore? I would expect the code to spawn multiple requests to get that metric data

Reply:
as suggested above, once you get a batch, spawn a few concurrent requests per to the allowed in datadog by the rate limit

Reply:
resolved now